libDF/src/tract.rs

```diff
@@ -0,0 +1,724 @@
+        reduce_mask: ReduceMask,
+    ) -> Self {
+        Self {
+            n_ch,
+            post_filter,
+            min_db_thresh,
+            max_db_erb_thresh,
+            max_db_df_thresh,
+            reduce_mask,
+        }
+    }
+}
+
+pub type TractModel = TypedSimpleState<TypedModel, TypedSimplePlan<TypedModel>>;
+
+pub struct DfTract {
+    enc: TractModel,
+    erb_dec: TractModel,
+    df_dec: TractModel,
+    pub df_lookahead: usize,
+    pub conv_lookahead: usize,
+    pub sr: usize,
+    pub ch: usize,
+    pub fft_size: usize,
+    pub hop_size: usize,
+    pub nb_erb: usize,
+    pub min_nb_erb_freqs: usize,
+    pub nb_df: usize,
+    pub n_freqs: usize,
+    pub df_order: usize,
+    pub post_filter: bool,
+    pub alpha: f32,
+    pub min_db_thresh: f32,
+    pub max_db_erb_thresh: f32,
+    pub max_db_df_thresh: f32,
+    pub reduce_mask: ReduceMask,
+    df_states: Vec<DFState>,
+    spec_buf: Tensor,
+    erb_buf: Tensor,
+    cplx_buf: Tensor,
+    m_zeros: Vec<f32>,
+    rolling_spec_buf: VecDeque<Tensor>,
+}
+
+impl DfTract {
+    pub fn new(dfp: DfParams, rp: &RuntimeParams) -> Result<Self> {
+        #[cfg(feature = "timings")]
+        let t0 = Instant::now();
+        let config = dfp.config;
+        let model_cfg = config.section(Some("deepfilternet")).unwrap();
+        let df_cfg = config.section(Some("df")).unwrap();
+        let ch = rp.n_ch;
+
+        let (enc, enc_delay) = init_encoder_from_read(&mut Cursor::new(dfp.enc), df_cfg, ch)?;
+        let (erb_dec, erb_dec_delay) =
+            init_erb_decoder_from_read(&mut Cursor::new(dfp.erb_dec), model_cfg, df_cfg, ch)?;
+        let (df_dec, df_dec_delay) =
+            init_df_decoder_from_read(&mut Cursor::new(dfp.df_dec), model_cfg, df_cfg, ch)?;
+        let enc = SimpleState::new(enc.into_runnable()?)?;
+        let erb_dec = SimpleState::new(erb_dec.into_runnable()?)?;
+        let df_dec = SimpleState::new(df_dec.into_runnable()?)?;
+        #[cfg(feature = "timings")]
+        let t1 = Instant::now();
+
+        let sr = df_cfg.get("sr").unwrap().parse::<usize>()?;
+        let hop_size = df_cfg.get("hop_size").unwrap().parse::<usize>()?;
+        let fft_size = df_cfg.get("fft_size").unwrap().parse::<usize>()?;
+        let min_nb_erb_freqs = df_cfg.get("min_nb_erb_freqs").unwrap().parse::<usize>()?;
+        let nb_erb = df_cfg.get("nb_erb").unwrap().parse::<usize>()?;
+        let nb_df = df_cfg.get("nb_df").unwrap().parse::<usize>()?;
+        let df_order = df_cfg
+            .get("df_order")
+            .unwrap_or_else(|| model_cfg.get("df_order").unwrap())
+            .parse::<usize>()?;
+        let df_lookahead = df_cfg
+            .get("df_lookahead")
+            .unwrap_or_else(|| model_cfg.get("df_lookahead").unwrap())
+            .parse::<usize>()?
+            .max(0);
+        // TODO: Why do I need 2 frames of delay here?
+        let conv_lookahead = model_cfg.get("conv_lookahead").unwrap().parse::<usize>()?.max(2);
```

Comment:
`max(2)` is needed so that the pulse-delay of the predicted mask matches the spectrogram frame

Reply:
DeepFilterNet2 model uses a lookahead of 2 frames so this does not matter.

Not sure why this is needed for models that are not trained with a lookahead.